code/layout-analyzer.js

```diff
@@ -130,22 +199,97 @@ window.addEventListener('DOMContentLoaded', () => {
     table.innerHTML = `<tr><td></td>${cols}<td></td></tr>`;
   };
 
-  // compute the same-finger usages and rolls
-  const computeDigrams = () => {
-    const skuCount = {}; // same-key usage
-    const sfuCount = {}; // same-finger usage
-    const skuDigrams = {};
-    const sfuDigrams = {};
-    const inwardDigrams = {};
-    const outwardDigrams = {};
-    const extendedRolls = {};
-    const scisors = {};
-
-    const fingers = ['l5', 'l4', 'l3', 'l2', 'r2', 'r3', 'r4', 'r5'];
-    fingers.forEach(finger => {
-      sfuCount[finger] = 0;
-      skuCount[finger] = 0;
-    });
+  // Returns a custom iterator, similar to Rust’s std::slice::Windows.
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_generators
+  // https://doc.rust-lang.org/std/primitive.slice.html#method.windows
+  const arrayWindows = (arr, len) => ({
+    *[Symbol.iterator]() {
+      for (let i = 0; i <= arr.length - len; i++) {
+        yield arr.slice(i, i + len);
+      }
+    }
+  });
+
+  const computeNGrams = () => {
+    const ngrams = Object.fromEntries([
+        // Digrams
+        'sfb',  // Same Finger Bigram
+        'skb',  // Same Key Bigram
+        'lsb',  // Lateral Strech Bigram
+        'handChange',  // Two keys typed by different hands
+        'scisor',  // Roll with uncomfortable height difference between the keys
+        'extendedScisor',  // scisor + lsb
+        'inwardRoll',   // Roll in the pinky -> index direction
+        'outwardRoll',  // Roll in the index -> pinky direction
+
+        // Trigrams
+        'redirect',  // Two rolls going in different directions
+        'badRedirect', // Redirect that doesn’t use the index
+        'sfs',  // Same Finger Skipgram (sfb with other key in the middle)
+        'sks',  // Same Key Skipgram (skb with other key in the middle)
+        'other',  // unused, is just two simplo digrams, nothing to note.
+      ].map(digramType => [digramType, {}])
+    );
```

Comment:
I think this table/enumeration is very important and should be more visible somehow. Maybe it could be a top-level constant (possibly at the top of the file). 

Aligning the comments vertically might make it a bit more readable.

Reply:
I’m not entirely sure how I could make it more visible. Making it a global variable seems pointless, since it’s never used elsewhere, and we’d need to reset the variable before writing to it. If you have an idea I’m all ears, though.

Reply:
I agree it won’t change how the code works, but putting this into a top-level global would achieve two goals : 
- give this enumeration a name ;
- make it impossible to miss.

With such an approach the line here would become : 
```js
const ngrams = Object.fromEntries(NGRAM_CATEGORIES).map(ngramType => [ngramType, {}]));
```

(NGRAM_CATEGORIES, NGRAM_CLASSES, whatever you think fits best)

Reply:
OOoooooh I actually really like this approach ! I’ll push another commit, give me a minute